jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -728,55 +736,55 @@ protected void writeHttpContent(Request request, Response response, Callback cal
 
     protected void putHeaders(Response response, HttpContent content, long contentLength)
     {
-        // TODO it is very inefficient to do many put's to a HttpFields, as each put is a full iteration.
-        //      it might be better remove headers en masse and then just add the extras:
-        // NOTE: If these headers come from a Servlet Filter we shouldn't override them here.
-//        headers.remove(EnumSet.of(
-//            HttpHeader.LAST_MODIFIED,
-//            HttpHeader.CONTENT_LENGTH,
-//            HttpHeader.CONTENT_TYPE,
-//            HttpHeader.CONTENT_ENCODING,
-//            HttpHeader.ETAG,
-//            HttpHeader.ACCEPT_RANGES,
-//            HttpHeader.CACHE_CONTROL
-//            ));
-//        HttpField lm = content.getLastModified();
-//        if (lm != null)
-//            headers.add(lm);
-//        etc.
+        HttpFields.Mutable headers = response.getHeaders();
```

Comment:
You removed the TODO but did not handle its concern. Are we giving up on those?

Reply:
You're talking about the bit of the TODO that speaks about not overriding the headers if they come from a servlet filter? I'm talking about the fact that there are many `add` and `put` calls and that each one is rather expensive. AFAICT this hasn't changed so maybe that TODO could be left in place.

Reply:
@lorban we are talking about the same thing.
line 755 removes 3 headers that are then added back, instead of 3 puts.

I could not do that with more headers, because the other headers have differing logic about existing headers, which I have attempted to explain in the comments.